File: legacy-runtime/src/main/java/com/xiaomi/network/WeightedHost.java
package com.xiaomi.network;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/network/WeightedHost.class */
class WeightedHost implements Comparable<WeightedHost> {
    private static final int HISTORY_SIZE = 30;
    private final LinkedList<AccessHistory> accessHistories;
    String host;
    private long touchedTime;
    protected int weight;

    public WeightedHost() {
        this(null, 0);
    }

    public WeightedHost(String str) {
        this(str, 0);
    }

    public WeightedHost(String str, int i) {
        this.accessHistories = new LinkedList<>();
        this.touchedTime = 0L;
        this.host = str;
        this.weight = i;
    }

    protected void addAccessHistory(AccessHistory accessHistory) {
        synchronized (this) {
            if (accessHistory != null) {
                this.accessHistories.add(accessHistory);
                int weight = accessHistory.getWeight();
                if (weight > 0) {
                    this.weight += accessHistory.getWeight();
                } else {
                    int i = 0;
                    for (int size = this.accessHistories.size() - 1; size >= 0 && this.accessHistories.get(size).getWeight() < 0; size--) {
                        i++;
                    }
                    this.weight += weight * i;
                }
                if (this.accessHistories.size() > 30) {
                    this.weight -= this.accessHistories.remove().getWeight();
                }
            }
        }
    }

    @Override // java.lang.Comparable
    public int compareTo(WeightedHost weightedHost) {
        if (weightedHost == null) {
            return 1;
        }
        return weightedHost.weight - this.weight;
    }

    public WeightedHost fromJSON(JSONObject jSONObject) throws JSONException {
        synchronized (this) {
            this.touchedTime = jSONObject.getLong("tt");
            this.weight = jSONObject.getInt("wt");
            this.host = jSONObject.getString("host");
            JSONArray jSONArray = jSONObject.getJSONArray("ah");
            for (int i = 0; i < jSONArray.length(); i++) {
                this.accessHistories.add(new AccessHistory().fromJSON(jSONArray.getJSONObject(i)));
            }
        }
        return this;
    }

    public ArrayList<AccessHistory> getAccessHistory() {
        ArrayList<AccessHistory> arrayList;
        synchronized (this) {
            arrayList = new ArrayList<>(this.accessHistories);
        }
        return arrayList;
    }

    public ArrayList<AccessHistory> getUnTouchedAccessHistory() {
        ArrayList<AccessHistory> arrayList;
        synchronized (this) {
            arrayList = new ArrayList<>();
            for (AccessHistory accessHistory : this.accessHistories) {
                if (accessHistory.getTime() > this.touchedTime) {
                    arrayList.add(accessHistory);
                }
            }
            this.touchedTime = System.currentTimeMillis();
        }
        return arrayList;
    }

    public JSONObject toJSON() throws JSONException {
        JSONObject jSONObject;
        synchronized (this) {
            jSONObject = new JSONObject();
            jSONObject.put("tt", this.touchedTime);
            jSONObject.put("wt", this.weight);
            jSONObject.put("host", this.host);
            JSONArray jSONArray = new JSONArray();
            Iterator<AccessHistory> it = this.accessHistories.iterator();
            while (it.hasNext()) {
                jSONArray.put(it.next().toJSON());
            }
            jSONObject.put("ah", jSONArray);
        }
        return jSONObject;
    }

    public String toString() {
        return this.host + ":" + this.weight;
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/network/WeightedHost.kt
package com.xiaomi.network

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

open class WeightedHost @JvmOverloads constructor(
    @JvmField var host: String? = null,
    @JvmField var weight: Int = 0
) : Comparable<WeightedHost> {
    
    private val accessHistories = LinkedList<AccessHistory>()
    private var touchedTime: Long = 0

    companion object {
        private const val HISTORY_SIZE = 30
    }

    internal fun addAccessHistory(accessHistory: AccessHistory?) {
        synchronized(this) {
            if (accessHistory != null) {
                accessHistories.add(accessHistory)
                val weightVal = accessHistory.weight
                if (weightVal > 0) {
                    weight += weightVal
                } else {
                    var i = 0
                    var index = accessHistories.size - 1
                    while (index >= 0 && accessHistories[index].weight < 0) {
                        i++
                        index--
                    }
                    weight += weightVal * i
                }
                if (accessHistories.size > HISTORY_SIZE) {
                    weight -= accessHistories.remove().weight
                }
            }
        }
    }

    override fun compareTo(other: WeightedHost): Int {
        return other.weight - this.weight
    }

    @Throws(JSONException::class)
    fun fromJSON(jSONObject: JSONObject): WeightedHost {
        synchronized(this) {
            touchedTime = jSONObject.getLong("tt")
            weight = jSONObject.getInt("wt")
            host = jSONObject.getString("host")
            val jSONArray = jSONObject.getJSONArray("ah")
            for (i in 0 until jSONArray.length()) {
                accessHistories.add(AccessHistory().fromJSON(jSONArray.getJSONObject(i)))
            }
        }
        return this
    }

    fun getAccessHistory(): ArrayList<AccessHistory> {
        synchronized(this) {
            return ArrayList(accessHistories)
        }
    }

    fun getUnTouchedAccessHistory(): ArrayList<AccessHistory> {
        synchronized(this) {
            val arrayList = ArrayList<AccessHistory>()
            for (accessHistory in accessHistories) {
                if (accessHistory.time > touchedTime) {
                    arrayList.add(accessHistory)
                }
            }
            touchedTime = System.currentTimeMillis()
            return arrayList
        }
    }

    @Throws(JSONException::class)
    fun toJSON(): JSONObject {
        synchronized(this) {
            val jSONObject = JSONObject()
            jSONObject.put("tt", touchedTime)
            jSONObject.put("wt", weight)
            jSONObject.put("host", host)
            val jSONArray = JSONArray()
            for (history in accessHistories) {
                jSONArray.put(history.toJSON())
            }
            jSONObject.put("ah", jSONArray)
            return jSONObject
        }
    }

    override fun toString(): String {
        return "$host:$weight"
    }
}
